pkg/gds/store/trtl/iterator.go

```diff
@@ -0,0 +1,281 @@
+package trtl
+
+import (
+	"context"
+	"errors"
+	"io"
+
+	"github.com/rs/zerolog/log"
+	trtlpb "github.com/trisacrypto/directory/pkg/trtl/pb/v1"
+	"github.com/trisacrypto/directory/pkg/utils/wire"
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+	"google.golang.org/protobuf/proto"
+)
+
+type vaspIterator struct {
+	trtlIterator
+}
+
+// trtlIterator is an interface that is implemented by both the trtlBatchIterator and
+// trtlStreamingIterator to iterate over values in the trtl store. The general workflow
+// is to instantiate the iterator with either NewTrtlBatchIterator or
+// NewTrtlStreamingIterator, call Next() to move the iterator to the next value, and if
+// Next() returns True, call Key() and/or Value() to retrieve the current key and/or
+// value. If Next() returns false, the iterator is already pointing to the last value.
+// The caller should check Error() to verify that no errors occurred during iteration
+// and Release() to clean up networking resources.
+type trtlIterator interface {
+	Next() bool
+	Prev() bool
+	Seek(key []byte) bool
+	Key() []byte
+	Value() []byte
+	Error() error
+	Release()
+}
+
+// trtlBatchIterator implements a batch iterator for the trtl store. Internally, the
+// key-value pairs are loaded from the server in paginated batches. This provides the
+// caller with materialized views of the data, and is best suited for operations such
+// as reading all of the objects in a namespace at once.
+type trtlBatchIterator struct {
+	client        trtlpb.TrtlClient
+	values        []*trtlpb.KVPair
+	index         int
+	nextPageToken string
+	namespace     string
+	err           error
+}
+
+func NewTrtlBatchIterator(client trtlpb.TrtlClient, namespace string) *trtlBatchIterator {
+	return &trtlBatchIterator{
+		client:    client,
+		namespace: namespace,
+		index:     -1,
+		values:    make([]*trtlpb.KVPair, 0),
+	}
+}
+
+func (i *trtlBatchIterator) Next() bool {
+	if i.err != nil {
+		return false
+	}
+
+	if i.index >= len(i.values)-1 && i.nextPageToken == "" {
+		// No more values from the iterator
+		return false
+	}
+
+	i.index++
+	if i.index > 0 && i.index < len(i.values) {
+		// We have already loaded the next value
+		return true
+	}
+
+	request := &trtlpb.IterRequest{
+		Namespace: i.namespace,
+	}
+	if i.index > 0 {
+		// We need to request the next page of values
+		request.Options = &trtlpb.Options{
+			PageToken: i.nextPageToken,
+		}
+	}
+
+	var reply *trtlpb.IterReply
+	var err error
+	ctx, cancel := withContext(context.Background())
+	defer cancel()
+
+	if reply, err = i.client.Iter(ctx, request); err != nil {
+		i.err = err
+		return false
+	}
+
+	// Add the new values to the slice
+	i.values = append(i.values, reply.Values...)
+	i.nextPageToken = reply.NextPageToken
+	return true
```

Comment:
There is a slight edge case here where no new values are returned. E.g. because there is no snapshot isolation for paginated requests, the previous request could have had a next page token, but then everything on the next page was deleted before that page could be fetched. 

```suggestion
	return i.index < len(i.values)
```